tests/zfs-tests/tests/functional/cli_root/zfs_destroy/zfs_destroy_017_pos.ksh

```diff
@@ -0,0 +1,289 @@
+# x tank
+# x ├── #bookmark
+# x ├── @snapshot
+# x ├── filesystem
+# x ├── filesystem-with-bookmark
+# x │   └── #bookmark
+# x ├── filesystem-with-snapshot
+# x │   └── @snapshot
+# x ├── filesystem-with-volume
+# x │    └── volume
+# x └── filesystem-with-snapshotted-volume
+# x     └── volume
+# x         └── @snapshot
+setup_hierarchy
+log_must zfs destroy -t all -r "$destroyroot/tank"
+log_mustnot zfs get name "$destroyroot/tank"
+log_mustnot zfs get name "$destroyroot/tank#bookmark"
+log_mustnot zfs get name "$destroyroot/tank@snapshot"
+log_mustnot zfs get name "$destroyroot/tank/filesystem"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-bookmark"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshot/volume"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume"
+
+# Verify boundaries of recursively deleting volumes and snapshots
+#   tank
+#   ├── #bookmark
+#   ├── @snapshot
+#   ├── filesystem
+#   ├── filesystem-with-bookmark
+#   │   └── #bookmark
+#   ├── filesystem-with-snapshot
+#   │   └── @snapshot
+#   ├── filesystem-with-volume
+#   │    └── volume
+#   └── filesystem-with-snapshotted-volume
+# x     └── volume
+# x         └── @snapshot
+setup_hierarchy
+log_must zfs destroy -t volume,snapshot -r \
+	 "$destroyroot/tank/filesystem-with-snapshotted-volume/volume"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume@snapshot"
+
+# Verify boundaries of recursively deleting filesystems and snapshots
+#   tank
+#   ├── #bookmark
+# x ├── @snapshot
+# x ├── filesystem
+#   ├── filesystem-with-bookmark
+#   │   └── #bookmark
+# x ├── filesystem-with-snapshot
+# x │   └── @snapshot
+#   ├── filesystem-with-volume
+#   │    └── volume
+#   └── filesystem-with-snapshotted-volume
+#       └── volume
+# x         └── @snapshot
+setup_hierarchy
+log_must zfs destroy -t filesystem,snapshot -r "$destroyroot/tank"
+log_must zfs get name "$destroyroot/tank"
+log_must zfs get name "$destroyroot/tank#bookmark"
+log_mustnot zfs get name "$destroyroot/tank@snapshot"
+log_mustnot zfs get name "$destroyroot/tank/filesystem"
+log_must zfs get name "$destroyroot/tank/filesystem-with-bookmark"
+log_must zfs get name "$destroyroot/tank/filesystem-with-bookmark#bookmark"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshot/volume"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume"
+log_must zfs get name "$destroyroot/tank/filesystem-with-volume"
+log_must zfs get name "$destroyroot/tank/filesystem-with-volume/volume"
+log_must zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume"
+log_must zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume"
+log_mustnot zfs get name "$destroyroot/tank/filesystem-with-snapshotted-volume/volume@snapshot"
```

Comment:
Destroying with `-t filesystem,snapshot` I would expect it to not destroy `tank` because it has a bookmark, nor `filesystem-with-volume/volume`, or `filesystem-with-snapshotted-volume/volume`. But, the test run is saying tank is deleted:

```
16:28:43.79 SUCCESS: zfs destroy -t filesystem,snapshot -r testpool/testfs1/tank
16:28:43.80 cannot open 'testpool/testfs1/tank': dataset does not exist
16:28:43.80 ERROR: zfs get name testpool/testfs1/tank exited 1
```

Does this seem wrong to anyone else?

Reply:
It is a little surprising to me. I confirmed bookmarks do not prevent deletion in FreeBSD's ZFS either. There don't seem to be any tests in ZoL to verify this behavior, nor do I see documentation describing what the expected behavior is.

@ahrens is this the intended behavior when destroying datasets with bookmarks?

Reply:
That is definitely one surprising bit. I'll look in to why the existence of a volume isn't preventing the destruction of its parent, likely tomorrow or Monday.